cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,793 @@
+  thrust::copy(handle.get_thrust_policy(), vertex_begin, vertex_end, ranks.begin());
+
+  thrust::for_each(
+    handle.get_thrust_policy(),
+    vertex_begin,
+    vertex_end,
+    [out_degrees = raft::device_span<edge_t const>(out_degrees.data(), out_degrees.size()),
+     ranks       = raft::device_span<vertex_t>(ranks.data(), ranks.size()),
+     v_first     = graph_view.local_vertex_partition_range_first()] __device__(auto v) {
+      auto v_offset = v - v_first;
+
+      if (out_degrees[v_offset] == 0) { ranks[v_offset] = invalid_vertex_id<vertex_t>::value; }
+    });
+
+  if (debug) { raft::print_device_vector("ranks: ", ranks.data(), ranks.size(), std::cout); }
+
+  size_t loop_counter                              = 0;
+  vertex_t nr_remaining_vertices_in_last_iteration = 0;
+  while (true) {
+    loop_counter++;
+    // if (debug) {
+    cudaDeviceSynchronize();
+    std::cout << "------------Mis loop, counter: ---------- " << loop_counter << std::endl;
+    // }
+    //
+    // Copy ranks into temporary vector to begin with
+    //
+    // Caches for ranks
+
+    rmm::device_uvector<vertex_t> temporary_ranks(local_vtx_partitoin_size, handle.get_stream());
+    thrust::copy(handle.get_thrust_policy(), ranks.begin(), ranks.end(), temporary_ranks.begin());
+
+    //
+    // Select a random set of candidate vertices
+    //
+    thrust::default_random_engine g;
+    g.seed(0);
+    thrust::shuffle(
+      handle.get_thrust_policy(), remaining_vertices.begin(), remaining_vertices.end(), g);
+
+    vertex_t nr_candidates =
+      (remaining_vertices.size() < 1024)
+        ? remaining_vertices.size()
+        : std::min(static_cast<vertex_t>((0.50 + 0.25 * loop_counter) * remaining_vertices.size()),
+                   static_cast<vertex_t>(remaining_vertices.size()));
+
+    // Set temporary ranks of non-candidate vertices to -Inf
+    thrust::for_each(handle.get_thrust_policy(),
+                     remaining_vertices.begin(),
+                     remaining_vertices.end() - nr_candidates,
+                     [temporary_ranks =
+                        raft::device_span<vertex_t>(temporary_ranks.data(), temporary_ranks.size()),
+                      v_first = graph_view.local_vertex_partition_range_first(),
+                      debug   = debug] __device__(auto v) {
+                       //
+                       // if rank of a non-candidate vertex is not +Inf (i.e. the vertex
+                       // is not already in MIS), set it to -Inf
+                       //
+                       auto v_offset = v - v_first;
+                       if (temporary_ranks[v_offset] < std::numeric_limits<vertex_t>::max()) {
+                         temporary_ranks[v_offset] = invalid_vertex_id<vertex_t>::value;
+                         if (debug)
+                           printf("Setting rank of %d to %d\n", v, temporary_ranks[v_offset]);
+                       }
+                     });
```

Comment:
And you can use thrust::replace_if.

https://thrust.github.io/doc/group__replacing_ga9953891234efd44b09942f7d6381ff79.html#ga9953891234efd44b09942f7d6381ff79